### ISSUE TITLE
mypy complains about monkey-patching classes in another module

### ISSUE THREAD

**Issue by timabbott** [python/mypy#1152]
This issue affects unannotated code.  Here's some example code from https://github.com/zulip/zulip/blob/master/zerver/lib/test_helpers.py#L47:

```
@contextmanager
def simulated_queue_client(client):
    real_SimpleQueueClient = queue_processors.SimpleQueueClient
    queue_processors.SimpleQueueClient = client
    yield
    queue_processors.SimpleQueueClient = real_SimpleQueueClient
```

(There's a similar issue with replacing a function rather than a class; and an example a few lines down in the Zulip codebase)

**Comment by JukkaL** [comment]
Can you give an example where replacing a function fails? Some kinds of function monkey patching should work, and it would be nice to have an example that is still broken.

In general, type checking the compatibility of classes is hard, but at least these should be fixed to work:

```
mod.Cls = cast(Any, mod.Cls2)

def f():  # no annotation -> don't complain!
    mod.Cls = mod.Cls2
```

This is a little trickier:

```
def f() -> None:
    mod.Cls = mod.Cls2   # When exactly should this be okay? When should it not be okay?
```

**Comment by timabbott** [comment]
Hmm I can't reproduce the function issue so let's remove that from this issue for now; I'll open another if I can reproduce.

**Comment by gvanrossum** [comment]
General suggestion: can you copy/paste the exact error message(s) you got from mypy? Those are often invaluable hints for finding the root of the problem (and figuring out how to run mypy on exactly the code you were using takes much longer).

**Comment by timabbott** [comment]
Oops, yeah, sorry about that, this is the error:
```
zerver/lib/test_helpers.py: note: In function "simulated_queue_client":
zerver/lib/test_helpers.py:57: error: Invalid assignment target
zerver/lib/test_helpers.py:59: error: Invalid assignment target
```

**Comment by JukkaL** [comment]
Here's a complete example which should probably not generate an error:

```py
# main.py
from typing import Any
import m

def f(x: Any) -> None:
    m.C = x  # Cannot assign to a type
```

```py
# m.py
class C: pass
```

**Comment by blueyed** [comment]
I've tried this method to change/enhance the types of attributes for a class from an external module this way.

```python
# m.py
class C: pass
```

```python
from typing import cast
import m

class TypedC:
    typed_attr: str

m.C = TypedC  # Cannot assign to a type
m.C = cast(TypedC, m.C)  # Cannot assign to a type

# works - a bit surprising (you might expect this to e.g. use `Any` then).
m.C = TypedC  # type: ignore
```

(Using `setattr` avoids the error, but does not use the new type then.)

**Comment by hauntsaninja** [comment]
I'm fine with mypy complaining about this. Monkeypatching classes is unsound given nominal typing.
